src/Disks/ObjectStorages/Encrypted/EncryptedObjectStorage.cpp

```diff
@@ -0,0 +1,198 @@
+#include "EncryptedObjectStorage.h"
+
+#include <filesystem>
+#include <Disks/IO/CachedOnDiskWriteBufferFromFile.h>
+#include <Disks/ObjectStorages/DiskObjectStorageCommon.h>
+#include <IO/BoundedReadBuffer.h>
+#include <IO/ReadBufferFromFileBase.h>
+#include <IO/WriteBufferFromEncryptedFile.h>
+#include <Interpreters/Context.h>
+#include <Common/CurrentThread.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int DATA_ENCRYPTION_ERROR;
+}
+
+String EncryptedObjectStorageSettings::findKeyByFingerprint(UInt128 key_fingerprint, const String & path_for_logs) const
+{
+    auto it = all_keys.find(key_fingerprint);
+    if (it == all_keys.end())
+    {
+        throw Exception(
+            ErrorCodes::DATA_ENCRYPTION_ERROR, "Not found an encryption key required to decipher file {}", quoteString(path_for_logs));
+    }
+    return it->second;
+}
+
+EncryptedObjectStorage::EncryptedObjectStorage(
+    ObjectStoragePtr object_storage_, EncryptedObjectStorageSettingsPtr enc_settings_, const std::string & enc_config_name_)
+    : object_storage(object_storage_), enc_settings(enc_settings_), enc_config_name(enc_config_name_), log(&Poco::Logger::get(getName()))
+{
+}
+
+DataSourceDescription EncryptedObjectStorage::getDataSourceDescription() const
+{
+    auto wrapped_object_storage_data_source = object_storage->getDataSourceDescription();
+    wrapped_object_storage_data_source.is_encrypted = true;
+    return wrapped_object_storage_data_source;
+}
+
+std::string EncryptedObjectStorage::generateBlobNameForPath(const std::string & path)
+{
+    return object_storage->generateBlobNameForPath(path);
+}
+
+ReadSettings EncryptedObjectStorage::patchSettings(const ReadSettings & read_settings) const
+{
+    ReadSettings modified_settings{read_settings};
+    modified_settings.encryption_settings = enc_settings;
+    return object_storage->patchSettings(modified_settings);
+}
+
+void EncryptedObjectStorage::startup()
+{
+    object_storage->startup();
+}
+
+bool EncryptedObjectStorage::exists(const StoredObject & object) const
+{
+    return object_storage->exists(object);
+}
+
+std::unique_ptr<ReadBufferFromFileBase> EncryptedObjectStorage::readObjects( /// NOLINT
+    const StoredObjects & objects,
+    const ReadSettings & read_settings,
+    std::optional<size_t> read_hint,
+    std::optional<size_t> file_size) const
+{
+    if (objects.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Received empty list of objects to read");
```

Comment:
There can be a valid case when we will have zero objects: https://github.com/ClickHouse/ClickHouse/issues/49480.